### PR TITLE
Migrated application form

### LINKED ISSUE #272: Migrate current admission page to Urán
Beyond the current system, this should include:

- a home page with the informations about the admission
- admitted students can be automatically made to collegists, so they won't have to register again

### LINKED ISSUE #631: Add programme (szak) to educational information
I just found out that we do not track it.

### PR DESCRIPTION
**Pull request by kdmnk**
Only the form, the committee pages will be another project.

Original page: felveteli.eotvos.elte.hu

### CODE REVIEW

File: app/Http/Livewire/ParentChildForm.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace App\Http\Livewire;
+
+use Livewire\Component;
+
+class ParentChildForm extends Component
```
Comment (kdmnk) [inline app/Http/Livewire/ParentChildForm.php:7]: A new general purpose Livewire component...

File: app/Http/Controllers/Auth/RegisterController.php [hunk 2]
```diff
@@ -62,12 +57,28 @@ public function showTenantRegistrationForm()
     {
         return view('auth.register', [
             'user_type' => Role::TENANT,
-            'faculties' => Faculty::all(),
-            'workshops' => Workshop::all(),
             'countries' => require base_path('countries.php'),
         ]);
     }
 
+    public const USER_RULES = [
```
Comment (kdmnk) [inline app/Http/Controllers/Auth/RegisterController.php:64]: Mainly restructure only.
The Educational Information are not handled by the Registration Controller from now on.

File: config/filesystems.php [hunk 3]
```diff
@@ -81,4 +81,13 @@
...
     ],
 
+    /*
+    * Symbolic links
+    */
+
+    'links' => [
+        public_path('avatars') => storage_path('app/avatars'),
+        public_path('uploads') => storage_path('app/uploads'),
+    ]
```
Comment (kdmnk) [inline config/filesystems.php:91]: Will need to run php artisan storage:link to take effect

File: resources/views/components/input/file.blade.php [hunk 4]
```diff
@@ -1,28 +1,18 @@
 {{-- Do not forget to add "enctype='multipart/form-data'" to the form! --}}
```
Comment (kdmnk) [inline resources/views/components/input/file.blade.php:2]: reformat only...

File: resources/views/welcome.blade.php [hunk 5]
```diff
@@ -55,28 +53,32 @@
...
-                    @lang('main.faster')</a><br class="mobile-break" />
-                <a href="#">
-                    @lang('main.brilliant')</a><br class="mobile-break" />
-                <a href="#">
-                    @lang('main.essential')</a><br class="mobile-break" />
-                <a href="#">
-                    @lang('main.modern')</a><br class="mobile-break" />
-                <a href="https://github.com/luksan47/mars">
-                    @lang('main.open')</a><br class="mobile-break" />
+                <a href="https://eotvos.elte.hu">@lang('general.mars_descr')</a>
```
Comment (kdmnk) [inline resources/views/welcome.blade.php:62]: new subtitle on welcome page

File: .env.example [hunk 6]
```diff
@@ -63,5 +63,8 @@ KKT=2000
 WORKSHOP_BALANCE_EXTERN=0.45
 WORKSHOP_BALANCE_RESIDENT=0.6
 
+APPLICATION_DEADLINE=
```
Comment (luksan47) [inline .env.example:66]: Shouldn't we set it?

File: .env.example [hunk 7]
```diff
@@ -63,5 +63,8 @@ KKT=2000
...
+APPLICATION_EXTENDED=false
```
Comment (luksan47) [inline .env.example:67]: We might want to start commenting these env vars :D

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+use Carbon\Carbon;
+use Illuminate\Http\Request;
+use Storage;
+
+class ApplicationController extends Controller
+{
+
+    public function showApplicationForm(Request $request)
+    {
+        App::setLocale('hu');
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: Isn't there a better place for this? Maybe a middleware?

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        $data = [
+            'workshops' => Workshop::all(),
+            'faculties' => Faculty::all(),
+            'deadline' => self::getApplicationDeadline(),
+            'deadline_extended' => self::isDeadlineExtended(),
+            'countries' => require base_path('countries.php'),
+            'user' => $request->user()
+        ];
+        switch ($request->page) {
+            case 'educational':
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: These could be constants

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    public function storeApplicationForm(Request $request)
+    {
+        $user = $request->user();
+
+        if($user->application && $user->application->status == ApplicationForm::STATUS_SUBMITTED) {
+            return redirect()->route('application')->with('error', 'Már véglegesítette a jelentkezését!');
+        }
+
+        switch ($request->page) {
+            case 'personal':
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: You could use those constants here

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                return view('auth.application.files', $data);
+            case 'submit':
+                return view('auth.application.submit', $data);
+            default:
+                return view('auth.application.personal', $data);
+        }
+
+    }
+
+    public function storeApplicationForm(Request $request)
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php:50]: When do we store the password and email?

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+                    'programs' => 'required|array',
+                    'programs.*' => 'nullable|string'
+                ]);
+                EducationalInformation::updateOrCreate(['user_id' => $user->id],[
+                    'year_of_graduation' => $request->year_of_graduation,
+                    'high_school' => $request->high_school,
+                    'neptun' => $request->neptun,
+                    'year_of_acceptance' => date('Y'),
+                    'email' => $request->educational_email,
+                    'program' => $request->programs,
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php:92]: Does this also store `programs.*`?

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+                ]);
+                if($request->status == 'resident') {
+                    $user->setResident();
+                }
+                if($request->status == 'extern') {
+                    $user->setExtern();
+                }
+                $user->workshops()->sync($request->workshop);
+                ApplicationForm::updateOrCreate(['user_id' => $user->id],[
+                    'question_1' => $request->question_1,
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php:118]: This could be more generic with a loop+function, but okay

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+                $request->validate([
+                    'file' => 'required|file|mimes:pdf,jpg,jpeg,png,gif,svg|max:2048',
+                    'name' => 'required|string|max:255',
+                ]);
+                $path = $request->file('file')->store('uploads');
+                $user->application()->firstOrCreate()->files()->create(['path' => $path, 'name' => $request->name]);
+                break;
+            case 'files.delete':
+                $request->validate([
+                    'id' => 'required|exists:files',
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php:134]: This cannot be other applicant's file, right?

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+                ]);
+
+                $file = $user->application->files()->find($request->id);
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: This can never throw null pointer exception, right?

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+            case 'questions':
+                $request->validate([
+                    'status' => 'nullable|in:extern,resident',
+                    'workshop' => 'array',
+                    'workshop.*' => 'exists:workshops,id',
+                ]);
+                if($request->status == 'resident') {
+                    $user->setResident();
+                }
+                if($request->status == 'extern') {
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: else if

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+                    'id' => 'required|exists:files',
+                ]);
+
+                $file = $user->application->files()->find($request->id);
+                if($file){
+                    Storage::delete($file->path);
+                    $file->delete();
+                }
+                break;
+            case 'files.profile':
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: Is this a profile picture? I could use a comment here

File: app/Http/Controllers/Auth/ApplicationController.php [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+                break;
+            default:
+                abort(404);
+                break;
+        }
+        return redirect()->back()->with('message', __('general.successful_modification'));
+    }
+
+
+    const DELIMETER = '|';
```
Comment (luksan47) [inline app/Http/Controllers/Auth/ApplicationController.php]: Is this used somehwere?

File: app/Http/Controllers/Auth/RegisterController.php (hunk 2 above)
Comment (luksan47) [reply app/Http/Controllers/Auth/RegisterController.php:64]: Does this also mean that collegists won't be able to register from now on?

File: app/Http/Livewire/ParentChildForm.php (hunk 1 above)
Comment (luksan47) [reply app/Http/Livewire/ParentChildForm.php:10]: Could you please add a comment what it's used for?

File: app/Models/ApplicationForm.php [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+        return true;
+    }
+
+    /*
+    |--------------------------------------------------------------------------
+    | Private functions
+    |--------------------------------------------------------------------------
+    */
+
+    private static function compressData($array)
```
Comment (luksan47) [inline app/Models/ApplicationForm.php]: Could you please add some comments here and below?

File: app/Models/EducationalInformation.php [hunk 20]
```diff
@@ -18,10 +18,60 @@ class EducationalInformation extends Model
...
+
+        return join(', ', $this->program);
+    }
+
+    public function setProgramAttribute($value)
+    {
+        $this->attributes['program'] = self::compressData($value);
+    }
+
+    private static function compressData($array)
```
Comment (luksan47) [inline app/Models/EducationalInformation.php]: This should be moved to some utilities file since it's used at multiple places.

File: app/Models/User.php [hunk 21]
```diff
@@ -281,6 +291,9 @@ public function hasRoleBase(string $roleName): bool
                 return true;
             }
         }
+        if ($this->hasRole($roleName)) {
```
Comment (luksan47) [inline app/Models/User.php:310]: Why is this added?

File: app/View/Components/Input/Textarea.php [hunk 22]
```diff
@@ -6,9 +6,12 @@
 
 class Textarea extends Input
 {
-    public function __construct($id, $locale = null, $text = null, $s = 12, $m = null, $l = null, $xl = null, $onlyInput = false)
+    public $helper;
+
+    public function __construct($id, $locale = null, $helper = null, $text = null, $s = 12, $m = null, $l = null, $xl = null, $onlyInput = false)
     {
         parent::__construct($id, $locale, $text, $s, $m, $l, $xl, $onlyInput);
+        $this->helper = $helper;
```
Comment (luksan47) [inline app/View/Components/Input/Textarea.php:20]: What's helper? Could you please add a comment?

File: database/migrations/2022_01_12_211223_create_applications_table.php [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+            $table->text('language_exam')->nullable();
+            $table->text('competition')->nullable();
+            $table->text('publication')->nullable();
+            $table->text('foreign_studies')->nullable();
+            $table->text('question_1')->nullable();
+            $table->text('question_2')->nullable();
+            $table->text('question_3')->nullable();
+            $table->text('question_4')->nullable();
+            $table->timestamps();
+        });
```
Comment (luksan47) [inline database/migrations/2022_01_12_211223_create_applications_table.php:32]: Are these values really all nullable? Eg. `high_school_address` is required in the validations.

**Review by luksan47 — APPROVED** [review]
LGTM with minor comments

File: app/Http/Controllers/Auth/RegisterController.php (hunk 2 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/RegisterController.php:64]: Yes. But we can trick this: register them as guest and then modify them.

File: app/Http/Controllers/Auth/ApplicationController.php (hunk 8 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/ApplicationController.php]: Do we want a new middleware just for this?

File: app/Http/Controllers/Auth/ApplicationController.php (hunk 11 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/ApplicationController.php:50]: They need to register first with their basic data. Then, the only page they can access is the application form.

File: app/Http/Controllers/Auth/ApplicationController.php (hunk 15 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/ApplicationController.php]: It is validated, yes.

File: app/Http/Controllers/Auth/RegisterController.php (hunk 2 above)
Comment (sszajbely) [reply app/Http/Controllers/Auth/RegisterController.php:64]: Can we wait with the merge for a few weeks? Some collegists who are not yet registered have just been asked to do so...

File: app/Http/Controllers/Auth/ApplicationController.php (hunk 14 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/ApplicationController.php:134]: I will make a policy

File: app/Http/Controllers/Auth/RegisterController.php (hunk 2 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/RegisterController.php:64]: I do not want to merge this without the application-handling/viewer possibilities.

File: app/Http/Controllers/Auth/ApplicationController.php (hunk 12 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/ApplicationController.php:92]: Yes, it is compressed into a single string value

File: app/Http/Controllers/Auth/ApplicationController.php (hunk 14 above)
Comment (kdmnk) [reply app/Http/Controllers/Auth/ApplicationController.php:134]: The path is searched only among the given user's application, so it will not be a problem

File: app/Models/User.php (hunk 21 above)
Comment (kdmnk) [reply app/Models/User.php:310]: The invariant that all the role's that can have object has an object is not true in the case where the applicant has not yet set the resident/extern status. Should not be a problem elsewhere, but I add a comment.

File: database/migrations/2022_01_12_211223_create_applications_table.php (hunk 23 above)
Comment (kdmnk) [reply database/migrations/2022_01_12_211223_create_applications_table.php:32]: The user can add these after the database entry is created. Existence is validated on submission.

**Comment by luksan47** [comment]
Please see https://github.com/luksan47/mars#important-notice
